includes/utils/class-amp-dom-document.php

```diff
@@ -0,0 +1,313 @@
+	 * a requirement for AMP.
+	 *
+	 * @var string
+	 */
+	private $original_encoding;
+
+	/**
+	 * Associative array of encoding mappings.
+	 *
+	 * Translates HTML charsets into encodings PHP can understand.
+	 *
+	 * @var string[]
+	 */
+	private $encoding_map = [
+		'latin-1' => 'iso-8859-1',
+	];
+
+	/**
+	 * Creates a new AMP_DOM_Document object
+	 *
+	 * @link  https://php.net/manual/domdocument.construct.php
+	 *
+	 * @param string $version  Optional. The version number of the document as part of the XML declaration.
+	 * @param string $encoding Optional. The encoding of the document as part of the XML declaration.
+	 */
+	public function __construct( $version = '', $encoding = null ) {
+		$this->original_encoding = (string) $encoding ?: self::UNKNOWN_ENCODING;
+		parent::__construct( $version ?: '1.0', self::AMP_ENCODING );
+	}
+
+	/**
+	 * Load HTML from a string.
+	 *
+	 * @link  https://php.net/manual/domdocument.loadhtml.php
+	 *
+	 * @param string     $source  The HTML string.
+	 * @param int|string $options Optional. Specify additional Libxml parameters.
+	 * @return bool true on success or false on failure.
+	 */
+	public function loadHTML( $source, $options = 0 ) {
+		$source = $this->maybe_add_head_or_body( $source );
+
+		$this->original_encoding = $this->detect_and_strip_encoding( $source );
+
+		if ( self::AMP_ENCODING !== strtolower( $this->original_encoding ) ) {
+			$source = $this->adapt_encoding( $source );
+		}
+
+		// Force-add http-equiv charset to make DOMDocument behave as it should.
+		// See: http://php.net/manual/en/domdocument.loadhtml.php#78243.
+		$source = str_replace(
+			'<head>',
+			'<head><meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">',
+			$source
+		);
+
+		$success = parent::loadHTML( $source, $options );
+
+		if ( $success ) {
+			$this->encoding = self::AMP_ENCODING;
+			$head           = $this->getElementsByTagName( 'head' )->item( 0 );
+			$head->removeChild( $head->firstChild );
+		}
+
+		return $success;
+	}
+
+	/**
+	 * Dumps the internal document into a string using HTML formatting.
+	 *
+	 * @link  https://php.net/manual/domdocument.savehtml.php
+	 *
+	 * @param DOMNode $node Optional. Parameter to output a subset of the document.
+	 * @return string The HTML, or false if an error occurred.
+	 */
+	public function saveHTML( DOMNode $node = null ) {
+		// Force-add http-equiv charset to make DOMDocument behave as it should.
+		// See: http://php.net/manual/en/domdocument.loadhtml.php#78243.
+		$head    = $this->getElementsByTagName( 'head' )->item( 0 );
+		$charset = AMP_DOM_Utils::create_node(
+			$this,
+			'meta',
+			[
+				'http-equiv' => 'content-type',
+				'content'    => 'text/html; charset=' . self::AMP_ENCODING,
+			]
+		);
+		$head->insertBefore( $charset, $head->firstChild );
+
+		return str_replace( '<meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">', '', parent::saveHTML( $node ) );
```

Comment:
This replacement key makes me a bit nervous, as it means we have have to rely on libxml always serializing attributes the same way. As long as we have tests for it, then I suppose nothing to worry about.

Reply:
It's a tag with two string-based attributes. I'm not sure how far off this could get.

However, I'll replace it with a `preg_replace()` to make it case-insensitive and "quote-insensitive".